sdk/tests/utilities/test_token_refresh.py

```diff
@@ -186,3 +194,388 @@ def force_refresh(refresh_token):
+            self.assertEqual(f"{refreshing_token}", "mock_access_token")
+
+    def test_retries_on_429_status_code_using_refresh_token(self):
+        """
+        Ensures that in the event of a 429 HTTP Status Code being returned when communicating with an identity
+        provider, the request is retried.
+        """
+        refreshing_token = RefreshingToken(api_configuration=self.config)
+
+        with patch("requests.post", side_effect=[
+            # Get initial access token
+            MockApiResponse(
+                json_data={
+                    "access_token": "mock_access_token",
+                    "refresh_token": "mock_refresh_token",
+                    "expires_in": 1  # Expires almost immediately
+                },
+                status_code=200
+            ),
+            # Return a 429 on the second attempt
+            MockApiResponse(
+                json_data={
+                    "error": "rate_limit",
+                    "error_description": "API rate limit exceeded."
+                },
+                status_code=429
+            ),
+            # Return a 200 on the third attempt
+            MockApiResponse(
+                json_data={
+                    "access_token": "mock_access_token_2",
+                    "refresh_token": "mock_refresh_token",
+                    "expires_in": 60
+                },
+                status_code=200
+            ),
+        ]):
+            # Ensure that we were able to get the first access token
+            self.assertEqual(f"{refreshing_token}", "mock_access_token")
+
+            sleep(1)  # Wait for initial token to expire
+
+            # Try and get access token again forcing refresh, if we can get it then retry was called
+            self.assertEqual(f"{refreshing_token}", "mock_access_token_2")
+
+    def test_does_not_retry_on_4xx_status_code_other_than_429(self):
+        """
+        Ensures that we do not retry on other common 4xx status codes such as 400 - Bad Request
+        """
+        refreshing_token = RefreshingToken(api_configuration=self.config)
+
+        with patch("requests.post", side_effect=[
+            # Return a 429 on the first attempt
+            MockApiResponse(
+                json_data={
+                    "error": "rate_limit",
+                    "error_description": "API rate limit exceeded."
+                },
+                status_code=429,
+                headers={}
+            ),
+            # Return a 400 on the second attempt
+            MockApiResponse(
+                json_data={
+                    "error": "invalid_grant",
+                    "error_description": "The refresh token is invalid or expired."
+                },
+                status_code=400
+            ),
+        ]):
+
+            # Ensure that a 400 is raised as an error and not retried
+            with self.assertRaises(ValueError):
+                self.force_refresh(refreshing_token)
```

Comment:
not clear to me what this code is doing (`force_refresh`), i would've thought you'd want to 
* assert error raised
* call to mock response for 400 == 1
* call to mock response for 429 == 0

Reply:
Yes fair enough, the initial retry call is redundant, just a single call to return a 400 now, also asserting on exception message too.

Reply:
Also asserting on call count.